### PR TITLE
[ADD] project_task_materials_stock: New module to create stock and 

### PR DESCRIPTION
**Pull request by sergio-teruel**
analytic moves.

This PR close #113

### CODE REVIEW

File: project_task_materials_stock/README.rst [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+Known issues / Roadmap
+======================
+
+* ...
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/{project_repo}/issues>`_.
```
Comment (pedrobaeza) [inline project_task_materials_stock/README.rst]: You have to replace `{project_repo}` with `project`

File: project_task_materials_stock/README.rst [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/{project_repo}/issues/new?body=module:%20{module_name}%0Aversion:%20{version}%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (pedrobaeza) [inline project_task_materials_stock/README.rst]: Replace placeholders by their values

File: project_task_materials_stock/README.rst [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+along with this program. If not, see <http://www.gnu.org/licenses/agpl-3.0-standalone.html>.
+
+
+Credits
+=======
+
+Contributors
+------------
+
+* Rafael Blasco <rafabn@antiun.com>
```
Comment (pedrobaeza) [inline project_task_materials_stock/README.rst]: As I designed this, I should be here also.

File: project_task_materials_stock/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Antiun Ingeniería S.L. - Sergio Teruel
+# (c) 2015 Antiun Ingeniería S.L. - Carlos Dauden
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    'name': 'Project Task Materials Stock',
+    'summary': 'Create stock and analytic moves from '
+               'record products spent in a Task',
+    'version': '1.0',
```
Comment (pedrobaeza) [inline project_task_materials_stock/__openerp__.py]: Use version 8.0.1.0.0

File: project_task_materials_stock/data/project_task_materials_data.xml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data noupdate="1">
+
+        <record id="analytic_journal_sale_materials" model="account.analytic.journal">
+            <field name="code">SALMAT</field>
+            <field name="name">Consumed Materials</field>
```
Comment (pedrobaeza) [inline project_task_materials_stock/data/project_task_materials_data.xml]: I would call it "Tasks Materials"

File: project_task_materials_stock/models/project.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+from openerp import models, fields, api
+from openerp.tools.float_utils import float_round
+
+
+class ProjectTaskType(models.Model):
+    _inherit = 'project.task.type'
+
+    consume_material = fields.Boolean(
+        string='Consume Material',
+        help="""If you select this option, when a task is in this state
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Don't use triple quota, but:

```
    help="If you select this option, when a task is in this state"
         "consumes materials associated"
```

Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Rephrase it to: "If you mark this check, when a task goes to this state, it will consume the associated materials"

File: project_task_materials_stock/models/project.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        consumes materials associated""")
+
+
+class Task(models.Model):
+    _inherit = "project.task"
+
+    def _get_stock_move(self):
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Call all compute methods with `compute` word, not `get` or `check`

File: project_task_materials_stock/models/project.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        comodel_name='account.analytic.line', string='Analytic Line')
+
+    def _prepare_stock_move(self):
+        product = self.product_id
+        res = {
+            'product_id': product.id,
+            'name': product.partner_ref,
+            'state': 'confirmed',
+            'product_uom': product.uom_id.id,
+            'product_uos': product.uos_id and product.uos_id.id or False,
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: If you are not putting the same UoM in both fields (uom and uos), you cannot put the same quantity without calculating the converted quantity. Please put here also `product.uom_id.id` for simplicity (you are not going to work with another UoM in this case I think). You can even mention this question in known issues: "You cannot use another UoM in tasks except the product stock one".

File: project_task_materials_stock/models/project.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+            'product_uom_qty': self.quantity,
+            'product_uos_qty': self.quantity,
+            'origin': self.task_id.name,
+            'location_id':
+            product.product_tmpl_id.property_stock_procurement.id,
+            'location_dest_id': self.env.ref(
+                'stock.stock_location_customers').id,
+        }
+        if product.uos_id and product.uom_id and \
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: hehe, I have say nothing in the previous comment. Here you do the conversion, so perfect in this way. But now I see that you should ask for which UoM to use, because maybe people want to work with another one than UoS.

File: project_task_materials_stock/models/project.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            'project_task_materials_stock.analytic_journal_sale_materials')
+        res = {
+            'name': self.task_id.name + ': ' + product.name,
+            'ref': self.task_id.name,
+            'product_id': product.id,
+            'journal_id': journal.id,
+            'unit_amount': self.quantity,
+            'account_id': self.task_id.project_id.analytic_account_id.id,
+            'to_invoice':
+            self.task_id.project_id.analytic_account_id.to_invoice.id,
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: This field is provided by _hr_timesheet_invoice_ module, so you shouldn't assume that exists. If you don't want to do a glue module, you can do 2 things:
- See if there's a `_prepare_*` or an onchange method on standard that puts these values via inheritance when an analytic account is passed.
- Use attrgetter method to bypass errors in case there's no such value.

File: project_task_materials_stock/README.rst (hunk 3 above)
Comment (dreispt) [inline project_task_materials_stock/README.rst]: I see your name 3 lines below :smile_cat:

Comment (pedrobaeza) [inline project_task_materials_stock/README.rst]: This has been added afterwards, and that's why the comment is not gone :wink:

File: project_task_materials_stock/README.rst [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+Usage
+=====
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/140/8.0
+
+Known issues / Roadmap
+======================
+
+* ...
```
Comment (dreispt) [inline project_task_materials_stock/README.rst]: Roadmap: add support for budgeted/expected materials.

File: project_task_materials_stock/models/project.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+            if move.state == 'assigned':
+                move.do_unreserve()
+            if move.state in ['waiting', 'confirmed', 'assigned']:
+                move.state = 'draft'
+            move.unlink()
+
+    @api.one
+    def write(self, vals):
+        res = super(Task, self).write(vals)
+        if 'stage_id' in vals:
```
Comment (dreispt) [inline project_task_materials_stock/models/project.py]: The logic in `super()` could change the Stage.
We probably should rebrowse the current records and then use `if rec.stage_id in vals:`

File: project_task_materials_stock/models/project.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+            if self.consume_material:
```
Comment (dreispt) [inline project_task_materials_stock/models/project.py]: I fell this is a little obfuscating. It's longer but clearer to use `if self.stage_id.consume_material`

File: project_task_materials_stock/models/project.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+        return res
+
+    @api.multi
+    def action_assign(self):
+        self.stock_move_ids.action_assign()
+
+    @api.multi
+    def action_done(self):
+        self.stock_move_ids.action_done()
+        pass
```
Comment (dreispt) [inline project_task_materials_stock/models/project.py]: These `action_*` methods are not used anymore.
What is expected for them? Can they simply be removed?

File: project_task_materials_stock/models/project.py [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+            'account.analytic.line')
+        journal = self.env.ref(
+            'project_task_materials_stock.analytic_journal_sale_materials')
+        res = {
+            'name': self.task_id.name + ': ' + product.name,
+            'ref': self.task_id.name,
+            'product_id': product.id,
+            'journal_id': journal.id,
+            'unit_amount': self.quantity,
+            'account_id': self.task_id.project_id.analytic_account_id.id,
```
Comment (dreispt) [inline project_task_materials_stock/models/project.py]: For "Service Desk" use cases, the Analytic Account comes from an analytic_id in the Task object:
Projects are "Service Teams", and Tasks are Work Orders for specific contracts.
If possible, I would like to factor that in here: check for the existence on an `analytic_id` field in the Task, and use it if it exists.

**Comment by dreispt** [comment]
Travis error is unrelated, so CI :ok:

File: project_task_materials_stock/__openerp__.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+    'category': "Project Management",
+    'author': 'Antiun Ingeniería S.L., '
+              'Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'depends': ['stock_account', 'project_task_materials'],
+    'data': [
+        'data/project_task_materials_data.xml',
+        'views/project_view.xml',
+    ],
+    'installable': True,
```
Comment (dreispt) [inline project_task_materials_stock/__openerp__.py:20]: Is `'auto_install': True,` a bad idea?

File: project_task_materials_stock/models/project.py (hunk 14 above)
Comment (carlos-incaser) [inline project_task_materials_stock/models/project.py]: This action_* methods are created as hooks, but I can remove this...

File: project_task_materials_stock/models/project.py (hunk 13 above)
Comment (carlosdauden) [inline project_task_materials_stock/models/project.py]: The aim of this is that the moves are only created or deleted when the stage of the task has been modified.
Otherwise moves would be created in each write.

File: project_task_materials_stock/models/project.py (hunk 14 above)
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: These methods are triggered by buttons, and they're intended for making the task to act as a "picking", where you have the check availability and transfer buttons to perform these operations for all the materials added. You should remove `pass` though.

File: project_task_materials_stock/README.rst (hunk 11 above)
Comment (carlosdauden) [inline project_task_materials_stock/README.rst]: Can you explain this support? :smile:

File: project_task_materials_stock/__openerp__.py (hunk 16 above)
Comment (carlos-incaser) [inline project_task_materials_stock/__openerp__.py]: I'm sorry. Copy/Paste error...

File: project_task_materials_stock/models/project.py (hunk 15 above)
Comment (carlosdauden) [inline project_task_materials_stock/models/project.py]: :+1:

File: project_task_materials_stock/__openerp__.py (hunk 16 above)
Comment (sergio-teruel) [inline project_task_materials_stock/__openerp__.py:20]: auto-install :+1:

File: project_task_materials_stock/models/project.py (hunk 13 above)
Comment (dreispt) [inline project_task_materials_stock/models/project.py]: I can see that, the point is that the related field shortcut makes it needlessly less clear to read.

File: project_task_materials_stock/models/project.py (hunk 14 above)
Comment (dreispt) [inline project_task_materials_stock/models/project.py]: I see now. I was confused by the method names, can we make them closer to what they are for?
- "action_assign" is for  "Check (Stock) Availability"
- "action_done" is for "Transfer Materials"

File: project_task_materials_stock/models/project.py (hunk 15 above)
Comment (dreispt) [inline project_task_materials_stock/models/project.py]: I was thinking about it, and I guess that the Timesheets integrated with Tasks surely have the same problem.
Given this, I'm wondering if the design approach for the Service Desk feature is the best one.
Probably this deserves a discussion on it's own...

File: project_task_materials_stock/models/project.py (hunk 13 above)
Comment (sergio-teruel) [inline project_task_materials_stock/models/project.py]: Yes...  of course :grimacing:

**Comment by rafaelbn** [comment]
Rebuilding for testing.

File: project_task_materials_stock/models/project.py [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+        string='Consume Material',
+        help="If you mark this check, when a task goes to this state,"
+             "it will consume the associated materials")
+
+
+class Task(models.Model):
+    _inherit = "project.task"
+
+    def _compute_stock_move(self):
+        move_ids = [line.stock_move_id.id for line in self.material_ids
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Use `self.mapped('material_ids.stock_move_id')`

File: project_task_materials_stock/models/project.py [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+                    if line.stock_move_id]
+        self.stock_move_ids = self.env['stock.move'].browse(move_ids)
+
+    def _compute_analytic_line(self):
+        line_ids = [line.analytic_line_id.id for line in self.material_ids
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Use `self.mapped('material_ids.analytic_line_id')`

File: project_task_materials_stock/models/project.py [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+        if not self.stock_move_ids:
+            self.stock_state = 'pending'
+        elif self.stock_move_ids.filtered(lambda r: r.state == 'confirmed'):
+            self.stock_state = 'confirmed'
+        elif self.stock_move_ids.filtered(lambda r: r.state == 'assigned'):
+            self.stock_state = 'assigned'
+        elif self.stock_move_ids.filtered(lambda r: r.state == 'done'):
+            self.stock_state = 'done'
+
+    stock_move_ids = fields.One2many(
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Shouldn't this be Many2many?

File: project_task_materials_stock/models/project.py [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+        comodel_name='account.analytic.line', compute='_compute_analytic_line',
+        string='Analytic Lines')
+    consume_material = fields.Boolean(related='stage_id.consume_material')
+    stock_state = fields.Selection(
+        [('pending', 'Pending'),
+         ('confirmed', 'Confirmed'),
+         ('assigned', 'Assigned'),
+         ('done', 'Done')], compute='_check_stock_state', string='Stock State')
+
+    @api.one
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Don't make it @api.one because is deprecated in v9. Use @api.multi and iterate over the recordset:

```
for record in self:
```

Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Or in this case, you can directly use mapped for iterating in all stock.move:

```
moves = self.mapped('stock_move_ids')
for move in moves:
```

File: project_task_materials_stock/models/project.py [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+    def unlink_stock_move(self):
+        for move in self.stock_move_ids:
+            if move.state == 'assigned':
+                move.do_unreserve()
+            if move.state in ['waiting', 'confirmed', 'assigned']:
+                move.state = 'draft'
+            move.unlink()
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Do this in one operation at the end: `moves.unlink()` (code with the new approach)

File: project_task_materials_stock/models/project.py [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+
+    @api.one
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Don't use @api.one also. And in this case, it's worst because your are forcing to serialize a write over multiple records.

File: project_task_materials_stock/models/project.py [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+                self.material_ids.create_stock_move()
+                self.material_ids.create_analytic_line()
+            else:
+                self.unlink_stock_move()
+                self.analytic_line_ids.unlink()
+        return res
+
+    @api.multi
+    def action_assign(self):
+        self.stock_move_ids.action_assign()
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: This will raise error when multiple records. Use instead:

```
self.mapped('stock_move_ids').action_assign()
```

File: project_task_materials_stock/models/project.py [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+
+    @api.multi
+    def action_done(self):
+        self.stock_move_ids.action_done()
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: The same

File: project_task_materials_stock/models/project.py [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+            'product_uos_qty': self.quantity,
+            'origin': self.task_id.name,
+            'location_id':
+            product.product_tmpl_id.property_stock_procurement.id,
+            'location_dest_id': self.env.ref(
+                'stock.stock_location_customers').id,
+        }
+        return res
+
+    @api.one
```
Comment (pedrobaeza) [inline project_task_materials_stock/models/project.py]: Don't use @api.one

**Comment by rafaelbn** [comment]
Tested: (you can check here: http://3116482-115-2d3873.runbot1.odoo-community.org)

1 - Add instruction un usage to explain that there is a new field named "Consume Material" in "Task stage" to indicate when you want to consume materials.

2 - Error. Not cosuming material.
- Create a product "motor" stockable and set in the inventory 10 unit, you have 10 motors.
- Create a task "repair car" in a project "car shop" (for example!)
- In task, Input 1 hour
- In task, input 1 motor
- Change stage in task which has stage marked as "consume material"
- You can check that products are not cosumed (they must be)

**Comment by rafaelbn** [comment]
@dreispt @pedrobaeza your comments have beed done and fixed, please review if we can remove label "needs fixing" and set "need review". Thanks

**Comment by pedrobaeza** [comment]
That's because move has waiting availability state. One thing you can do is to raise a Warning if the material can be consumed due to it hasn't been reserved.

**Comment by sergio-teruel** [comment]
@rafaelbn @pedrobaeza Fixed sotck location.

File: project_task_materials_stock/README.rst [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+
+Configuration
+=============
+
+Check option 'Consume Material' in Task Stage to generate a stock move when
+the task is in that stage.
+
+Usage
+=====
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
```
Comment (rafaelbn) [inline project_task_materials_stock/README.rst:25]: Add instruction un usage to explain that there is a new field named "Consume Material" in "Task stage" to indicate when you want to consume materials

Comment (sergio-teruel) [inline project_task_materials_stock/README.rst:25]: @rafaelbn Change done.

**Comment by pedrobaeza** [comment]
Please rebase to get green tests

**Comment by sergio-teruel** [comment]
@pedrobaeza Rebase done

File: project_task_materials_stock/README.rst [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+
+This module extends project_task_material module to consume products spent in 
+a task and create analytic lines to manage costs and create invoices.  
+
+Configuration
+=============
+
+Check option 'Consume Material' in Task Stage to generate a stock move when
+the task is in that stage.
+
```
Comment (rafaelbn) [inline project_task_materials_stock/README.rst:19]: - Go to Setting -> Configuration - > Projects and enable option "Manage time estimation on tasks"

**Comment by rafaelbn** [comment]
Tested in runbot :+1: , please change readme with comment and check travis, thanks.

**Comment by sergio-teruel** [comment]
@rafaelbn changes done..

**Comment by rafaelbn** [comment]
In my opinion this module is ready to be merged but we need more reviews. Thanks

**Comment by ingenieriacloud** [comment]
:+1:  functional

**Comment by janetrp** [comment]
tested in runbot :+1:

**Comment by dreispt** [comment]
Congratulations! It got through the review pinhole :smile:
